nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -329,6 +340,28 @@ BtNavigator::navigateToPose()
   }
 }
 
+bool
+BtNavigator::goalCheckerCallback(std::shared_ptr<const typename Action::Goal> goal)
```

Comment:
It seems like this same logic could just be placed in the preemption checks https://github.com/ros-planning/navigation2/blob/7bc04a6c5f0f081292d04339ec1fcbd6d61bf669/nav2_bt_navigator/src/bt_navigator.cpp#L294 that we just blindly accept right now. We could just add a check there and modulate the return code appropriately.

Reply:
yes. But this implementation is a little bit higher in the callstack. Preemption checks come right after the `action-goal` is accepted

Reply:
But wouldn't adding 

```
if is_preempt_requested:
  // get and check goal's validity per specific server needs
  if OK:
    accept
  else:
    reject
```

Accomplish the same thing without modifying simple action server and enable the behavior you were looking to add to the BT navigator?

Reply:
nope.

I want to check if the xml is loadable before accepting the action_goal.

is_preempt_requested is called by the current active / running "goal", actually by the BT. So far more down the "callstack":

https://github.com/ros-planning/navigation2/blob/7bc04a6c5f0f081292d04339ec1fcbd6d61bf669/nav2_bt_navigator/src/bt_navigator.cpp#L320

So if the robot is just in IDLE and waiting for a new goal -> I would not be able to "pre" check if the xml is loadable with the `is_preempt_requested`. And this started questioning me if it is necessary to check this on multiple occasions:

1. in the `load_bt()` -> load the new bt
2. in `is_preempt_requested` (otherwise we would stop a current goal for a bad one)
3. Potentially some other places

This however is the point of minimal overlap and should be the exact place of this new (re-introduced) callback.

Also: You cannot REJECT it any more at this place as you accepted it per Action-server-design. So you would cancel it. But I think that would not be the worst, as I just got my head around that there could be multiple accepted goals (current + pending or potential more if wanted). So Canceling in this case would not cancel all goals (including the current, which I was mainly afraid of)


Still: I think it would be better if goals are REJECTED and not Cancelled. This would be a better implementation.

--------
(Probably OT) Also:
https://github.com/ros-planning/navigation2/blob/199ef58f04bce9da4fcf8800630d171665349015/nav2_util/include/nav2_util/simple_action_server.hpp#L230

This is the reason why with PREEMTION no new xmls can be loaded.. As it just updates the goal variable and the BT keeps running.
(No new `work()` gets triggered/launched)